web-lib-funcs.pl

```diff
@@ -12118,18 +12118,61 @@ sub delvar
 return &globals('delete', $variable, $scope);
 }
 
-# webmin_user_is_admin([username])
-# Returns 1 if the given user should be considered fully trusted
-sub webmin_user_is_admin
+# webmin_user_can_rpc()
+# Returns 1 if the given user can make remote calls
+sub webmin_user_can_rpc
 {
-my ($user) = @_;
-$user ||= $base_remote_user;
-my %access = &get_module_acl($user, "");
+my %access = &get_module_acl($base_remote_user, "");
 return 1 if ($access{'rpc'} == 1);	# Can make arbitary RPC calls
 return 0 if ($access{'rpc'} == 0);	# Cannot make RPCs
+}
+
+# webmin_user_login_mode()
+# Returns currently logged in user mode
+sub webmin_user_login_mode
+{
+# Default mode
+my $mode = 'root';
+my $prod = &get_product_name();
+
+# Check for foreign modules
+my $foreign_virtual_server
+    = &foreign_available("virtual-server");
+&foreign_require("virtual-server")
+    if ($foreign_virtual_server);
+my $foreign_server_manager
+    = &foreign_available("server-manager");
+&foreign_require("server-manager")
+    if ($foreign_server_manager);
+
+# Get current user and base user global permissions
+my %uaccess = &get_module_acl($remote_user, "");
+my %access = &get_module_acl($base_remote_user, "");
+
+# Check if mode must be restricted
+if ($uaccess{'_safe'} == 1 || $access{'_safe'} == 1 ||
+    $uaccess{'rpc'} == 0 || $access{'rpc'} == 0 ||
+    $prod ne "webmin") {
+        $mode = 'user';
```

Comment:
I don't think we can just rely in the _safe field, as this was introduced relatively recently - many users have less-privileged Webmin logins that weren't created as "safe users", but should not be considered root-equivalent.

Reply:
It would still work, and be set to `$mode=user` because of `$access{'rpc'} == 0` test?

Reply:
> I don't think we can just rely in the _safe field, as this was introduced relatively recently

I have checked deeper and it appears that all main functionality for the Safe Webmin user was added with [this](https://github.com/webmin/webmin/commit/2075fffe70a8c593d5045f76993793fe5c511356) commit. It already includes `_safe` flag.

Also, as I mentioned before, I expect that `$access{'rpc'} == 0` would also force a regular user mode, even if `_safe` flag is missing.

Reply:
So is there any value in checking for the `_safe` flag then?

Reply:
> So is there any value in checking for the _safe flag then?

Yeah, I test if `$access{'_safe'} == 1` or `$access{'rpc'} == 0` and if true it will fall to safe user mode.